### PR TITLE
VB-1189 scheduled task to delete reserved bookings is run for every instance

### PR DESCRIPTION
**Pull request by dhirajshettymoj**
## What does this pull request do?

Adds support for shedlock.
## What is the intent behind these changes?

Adding shedlock will introduce a lock mechanism to ensure only 1 instance is running the delete expired visits task.

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/config/SchedulerConfiguration.kt [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+import org.springframework.context.annotation.Configuration
+import org.springframework.context.annotation.Profile
+import org.springframework.scheduling.annotation.EnableScheduling
+import javax.sql.DataSource
+
+@Configuration
+@EnableScheduling
+@Profile("!test") // prevent scheduler running during integration tests
+@EnableSchedulerLock(
+  defaultLockAtLeastFor = "PT10M",
```
Comment (AledWEvans) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/config/SchedulerConfiguration.kt:16]: should this me less is 10 mins too much

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/task/VisitTask.kt [hunk 2]
```diff
@@ -18,6 +19,10 @@ class VisitTask(
   }
 
   @Scheduled(cron = "\${task.expired-visit.cron:0 0/15 * * * ?}")
+  @SchedulerLock(
+    name = "deleteExpiredVisitsTask",
+    lockAtLeastFor = "PT5M", lockAtMostFor = "PT10M"
```
Comment (AledWEvans) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/task/VisitTask.kt:24]: is 10mins too much,... might be reading this incorrectly

File: src/main/resources/db/migration/V2_07__add_shedlock_tables.sql [hunk 3]
```diff
@@ -0,0 +1,7 @@
+CREATE TABLE shedlock (
```
Comment (AledWEvans) [inline src/main/resources/db/migration/V2_07__add_shedlock_tables.sql:1]: shed_lock may be

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/task/VisitTask.kt (hunk 2 above)
Comment (dhirajshettymoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/task/VisitTask.kt:24]: Should I reduce it to 5?

File: src/main/resources/db/migration/V2_07__add_shedlock_tables.sql (hunk 3 above)
Comment (dhirajshettymoj) [reply src/main/resources/db/migration/V2_07__add_shedlock_tables.sql:1]: Yes, needs to be shedlock.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/config/SchedulerConfiguration.kt (hunk 1 above)
Comment (AledWEvans) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/config/SchedulerConfiguration.kt:16]: yep 5 mins fine I think

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/task/VisitTask.kt (hunk 2 above)
Comment (AledWEvans) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/task/VisitTask.kt:24]: sure

File: src/main/resources/db/migration/V2_07__add_shedlock_tables.sql (hunk 3 above)
Comment (AledWEvans) [reply src/main/resources/db/migration/V2_07__add_shedlock_tables.sql:1]: I had a look could not find a  way to change the table name, without hassel

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/config/SchedulerConfiguration.kt (hunk 1 above)
Comment (dhirajshettymoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/config/SchedulerConfiguration.kt]: @AledWEvans  I have now updated the lock (atLeastFor and AtMostFor) to 5 mins.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/task/VisitTask.kt (hunk 2 above)
Comment (dhirajshettymoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/task/VisitTask.kt]: I have now updated the lock (atLeastFor and AtMostFor) to 5 mins.

**Review by AledWEvans — APPROVED** [review]
Nice work LGTM

**Review by AledWEvans — APPROVED** [review]
LGTM, excellent testing code update ;-)

**Comment by dhirajshettymoj** [comment]
> LGTM, excellent testing code update ;-)

:) thanks for the code

**Pull request merged**